src/collections/decodeableMap.ts

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright (c) 2023, salesforce.com, inc.
+ * All rights reserved.
+ * Licensed under the BSD 3-Clause license.
+ * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+
+/**
+ * This is an extension of the Map class that can match keys whether they are encoded or decoded.
+ * Decoding the key can solve some edge cases in component fullNames such as Layouts and Profiles.
+ * See: https://github.com/forcedotcom/cli/issues/1683
+ *
+ * Examples:
+ *
+ * Given a map with entries:
+ * ```javascript
+ * 'layout#Layout__Broker__c-v1.1 Broker Layout' : {...}
+ * 'layout#Layout__Broker__c-v9%2E2 Broker Layout' : {...}
+ * ```
+ *
+ * `decodeableMap.has('layout#Layout__Broker__c-v1%2E1 Broker Layout')` --> returns `true`
+ * `decodeableMap.has('layout#Layout__Broker__c-v9.2 Broker Layout')` --> returns `true`
+ */
+export class DecodeableMap<K extends string, V> extends Map<string, V> {
+  // Internal map of decoded keys to encoded keys.
+  // E.g., { 'foo-v1.3 Layout': 'foo-v1%2E3 Layout' }
+  // This is initialized in the `keysMap` getter.
+  private internalkeysMap!: Map<string, string>;
```

Comment:
Good question.  I originally had the Map initialization done during declaration and it would fail during construction due to the way `super` works.  `super` must be the first call in a constructor, which means the map init was done after that.  So if you construct a new map by passing in a nested array, it will call `set()` which needs the internal map but it wasn't initialized until later.  To solve this problem I used the getter and conditionally initialized it there.
I'll add a comment.